### ISSUE TITLE
parsing message extensions inside a message body

### ISSUE THREAD

**Issue by mmmries** [tomas-abrahamsson/gpb#42]
I get an error when trying to parse protobuf definitions like:

```protobuf
package atlas.dispatch;

// Base resource message
message Shard {
  extensions 200 to max;

  // Column fields
  optional string guid             = 1;
  optional string name             = 2;

  // Extension fields
  extend Shard {
    optional int64 status_code = 200;
  }
}
```

The same message definition parses fine if you move the `extend` outside the message definition like this:

```protobuf
package atlas.dispatch;

// Base resource message
message Shard {
  extensions 200 to max;

  // Column fields
  optional string guid             = 1;
  optional string name             = 2;  
}

// Extension fields
extend Shard {
  optional int64 status_code = 200;
}
```

Google's `protoc` compiler handles both of those forms just fine. Is this something that gpb can easily support? I would be happy to help out with a pull request if you are interested in supporting this feature and you can point me in the right general direction.

**Comment by tomas-abrahamsson** [comment]
Hi, this seems like a useful thing to be able to do, given that extensions are already parsed, albeit obviously not at all possible locations. Pull requests are of course welcome, the `gpb_parse.yrl` file would need to be updated. the `msg_elem` would need to be extended to parse `extend` inside a message. Then, the `flatten_qualify_defnames/1` function would have to handle it too, I think. Possibly also `resolve_refs/1` and probably also `extend_msgs/1`.

To get some understanding of the data flow, the top-functions are `post_process_one_file/2` called for each file, then `post_process_all_files/2` once all imports have been parsed and post processed individually.

There are some unit tests in test/gpb_parse_tests.erl which might be useful to have a look at.

**Comment by mmmries** [comment]
:+1: Thanks for the quick intro. I'll take a run at the problem and see how far I can get.

**Comment by tomas-abrahamsson** [comment]
Thanks for contributing! Included in 3.18.9
